flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/runtime/operators/join/window/WindowJoinOperator.java

```diff
@@ -0,0 +1,589 @@
+import java.util.List;
+
+import static org.apache.flink.table.runtime.util.TimeWindowUtil.isWindowFired;
+
+/**
+ * Streaming window join operator.
+ *
+ * <p>Note: currently, {@link WindowJoinOperator} doesn't support early-fire and late-arrival. Thus
+ * late elements (elements belong to emitted windows) will be simply dropped.
+ *
+ * <p>Note: currently, {@link WindowJoinOperator} doesn't support DELETE or UPDATE_BEFORE input row.
+ */
+public abstract class WindowJoinOperator extends TableStreamOperator<RowData>
+        implements TwoInputStreamOperator<RowData, RowData, RowData>,
+                Triggerable<RowData, Long>,
+                KeyContext {
+
+    private static final long serialVersionUID = 1L;
+
+    private static final String LEFT_LATE_ELEMENTS_DROPPED_METRIC_NAME =
+            "leftNumLateRecordsDropped";
+    private static final String LEFT_LATE_ELEMENTS_DROPPED_RATE_METRIC_NAME =
+            "leftLateRecordsDroppedRate";
+    private static final String RIGHT_LATE_ELEMENTS_DROPPED_METRIC_NAME =
+            "rightNumLateRecordsDropped";
+    private static final String RIGHT_LATE_ELEMENTS_DROPPED_RATE_METRIC_NAME =
+            "rightLateRecordsDroppedRate";
+    private static final String WATERMARK_LATENCY_METRIC_NAME = "watermarkLatency";
+    private static final String LEFT_RECORDS_STATE_NAME = "left-records";
+    private static final String RIGHT_RECORDS_STATE_NAME = "right-records";
+
+    protected final InternalTypeInfo<RowData> leftType;
+    protected final InternalTypeInfo<RowData> rightType;
+    private final GeneratedJoinCondition generatedJoinCondition;
+
+    private final int leftWindowEndIndex;
+    private final int rightWindowEndIndex;
+
+    private final boolean[] filterNullKeys;
+    private final ZoneId shiftTimeZone;
+
+    /** Flag to prevent duplicate function.close() calls in close() and dispose(). */
+    private transient boolean functionsClosed = false;
+
+    private transient WindowTimerService<Long> windowTimerService;
+
+    // ------------------------------------------------------------------------
+    protected transient JoinConditionWithNullFilters joinCondition;
+
+    /** This is used for emitting elements with a given timestamp. */
+    protected transient TimestampedCollector<RowData> collector;
+
+    private transient WindowListState<Long> leftWindowState;
+    private transient WindowListState<Long> rightWindowState;
+
+    // ------------------------------------------------------------------------
+    // Metrics
+    // ------------------------------------------------------------------------
+
+    private transient Counter leftNumLateRecordsDropped;
+    private transient Meter leftLateRecordsDroppedRate;
+    private transient Counter rightNumLateRecordsDropped;
+    private transient Meter rightLateRecordsDroppedRate;
+    private transient Gauge<Long> watermarkLatency;
+
+    WindowJoinOperator(
+            InternalTypeInfo<RowData> leftType,
```

Comment:
I think pass serializer `TypeSerializer<RowData>` is better. Just like `AbstractWindowAggProcessor`.

Reply:
Could you please explain why `TypeSerializer<RowData>` is better?
I use `InternalTypeInfo ` because it is easier to create state description and create null padding row based on `arity` .
BTW, if use `TypeSerializer<RowData>` has big advantage, I could use `RowDataSerializer` instead of `InternalTypeInfo<RowData>`.

Reply:
The core requirement is `TypeSerializer<RowData>`, we just need it is OK.
Personally, I don't like `InternalTypeInfo`, TypeInformation contains lots of useless information.